devices/src/virtio/mmio.rs

```diff
@@ -171,6 +170,75 @@ impl MmioDevice {
             false
         }
     }
+
+    fn update_queue_field<F: FnOnce(&mut Queue)>(&mut self, f: F) {
+        if self.check_driver_status(DEVICE_FEATURES_OK, DEVICE_DRIVER_OK | DEVICE_FAILED) {
+            self.with_queue_mut(f);
+        } else {
+            #[cfg(not(release))]
+            warn!(
+                "update virtio queue in invalid state 0x{:x}",
+                self.driver_status
+            );
+        }
+    }
+
+    /// Update driver status according to the state machine defined by VirtIO Spec 1.0.
+    /// Please refer to VirtIO Spec 1.0, section 2.1.1 and 3.1.1.
+    ///
+    /// The driver MUST update device status, setting bits to indicate the completed steps
+    /// of the driver initialization sequence specified in 3.1. The driver MUST NOT clear
+    /// a device status bit. If the driver sets the FAILED bit, the driver MUST later reset
+    /// the device before attempting to re-initialize.
+    fn update_driver_status(&mut self, v: u32) {
+        // match changed bits
+        match !self.driver_status & v {
+            DEVICE_ACKNOWLEDGE if self.driver_status == 0 => {
+                self.driver_status = v;
+            }
+            DEVICE_DRIVER if self.driver_status == DEVICE_ACKNOWLEDGE => {
+                self.driver_status = v;
+            }
+            DEVICE_FEATURES_OK if self.driver_status == (DEVICE_ACKNOWLEDGE | DEVICE_DRIVER) => {
+                self.driver_status = v;
+            }
+            DEVICE_DRIVER_OK
+                if self.driver_status
+                    == (DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_FEATURES_OK) =>
+            {
+                self.driver_status = v;
+                if !self.device_activated && self.are_queues_valid() {
```

Comment:
nit: add a comment here:

```
// If the driver incorrectly sets up the queue, the following check will fail and take the device into an unusable state.
```

I think it's important to call out this scenario since it's something we can't currently recover from.

Reply:
done

Reply:
done